### PR TITLE
Add a section "Tips for designing types"

### PR DESCRIPTION
**Pull request by tkf**
This probably is too opinionated to be in the official documentation, but I thought is worth a shot.  What do you think?

If you want to look at a semi-rendered version, see:
https://github.com/tkf/ConstructionBase.jl/blob/tips/docs/src/index.md#tips-for-designing-typesid-type-tips

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -3,8 +3,120 @@
...
+```jldoctest tips
+abstract type OutputBy end
+struct Mutating <: OutputBy end
+struct Returning <: OutputBy end
+
+struct Add{O <: OutputBy, T}
+    outputby::O
+    value::T
+end
+
```
Comment (jw3126) [inline docs/src/index.md]: I like this trick very much and it did not occur to me before. I also like that it lets you manipulate the `O` via `setproperties`.

File: docs/src/index.md [hunk 2]
```diff
@@ -3,8 +3,120 @@
...
 ```@docs
 ConstructionBase
 ConstructionBase.constructorof
 ConstructionBase.setproperties
 ```
+
+## [Tips for designing types](@id type-tips)
+
+When designing types from scratch, it is recommended to structure the types in
```
[the remark this answers is not in the document]
Comment (jw3126) [reply docs/src/index.md]: or rather "often possible"

[the remark this answers is not in the document]
Comment (tkf) [reply docs/src/index.md]: "often possible" sounds good!

File: docs/src/index.md (hunk 1 above)
Comment (tkf) [reply docs/src/index.md]: I guess I was thinking backward:  Given Setfield, how can I make my types better? :smile:

File: docs/src/index.md [hunk 3]
```diff
@@ -13,14 +13,15 @@ ConstructionBase.setproperties
...
-such a way that overloading `constructorof` or `setproperties` is unnecessary in
-the first place.  For simple `struct`s whose type parameters can be determined
-from field values, `ConstructionBase` works without any customization, provided
-that the "type-less" constructor exists.  However, it is often useful or
-required to have type parameters that cannot be determined from field values.
-One way to solve this problem is to define singleton types that would determine
-the type parameters:
+When designing types from scratch, it is often possible to structure the types
+in such a way that overloading `constructorof` or `setproperties` is unnecessary
+in the first place.  It let types in your package work nicely with the ecosystem
```
Comment (jw3126) [inline docs/src/index.md]: That is a good point!

**Pull request merged**